### DISCUSSION TITLE
[Feature Request]: A way to stop all scheduled jobs, but wait for running jobs to complete

### DISCUSSION THREAD

**Discussion by jahands** [Hexagon/croner#244] [Ideas]
Maybe this is possible and I just can't find it in the docs. I'd like to handle SIGTERM gracefully like this:
1. Receive SIGTERM
2. Cancel all future cron runs
3. Wait for any jobs that are currently running to complete
4. Exit application

It would be cool if there was something like this:
`await scheduledJobs.drain()` or something like that

What's the recommended way to do this?

**Comment by Hexagon** [comment]
A graceful shutdown is carried out by iterating over scheduledJobs, running `.stop()` on each of the Jobs.

`.stop()` prevents and further execution and stops the scheduler loop, but does not forcefully stop any running executions. It also does remove the job from scheduledJobs. 

Note that jobs are only stored automatically in the scheduledJobs array if they are named.

**Reply by anshdivu** [reply]
Hi @Hexagon, I noticed the comment on the `stop()` function mentions that it "forcefully stops execution".
https://github.com/Hexagon/croner/blob/c7e2a8359ddf9303c3aff31aada7de5708bdd133/src/croner.ts#L290-L297

Looking at the implementation, it seems the function mostly clears the `setTimeout` to prevent future executions. I'm wondering if there might be something else in the codebase that contributes to a forceful stop, or if the comment might no longer be accurate?

Could you please clarify this?

**Comment by jahands** [comment]
Ah! I see - and the program doesn't terminate until the job is done running - that works!
I was kind of expecting to need a promise that delays SIGTERM signals until all crons are done, but it seems it handles that already.

Would still be nice, I think, to have a stopAndWait() method that returns a promise which resolves when the current execution (if any) is done

**Comment by warren-gallagher** [comment]
Iterating over the scheduledJobs is a bit tricky because the array is being modified as you are iterating thus:

```
for( const job of scheduledJobs ) {
  job.stop();
}
```
will skip some of the jobs. whereas:
```
while( scheduledJobs.length > 0 ) {
  job = scheduledJobs[0];
  job.stop();
}
```
will work.
